### PR TITLE
feat: add API entrypoint (#721)

### LINKED ISSUE #721: Implement API Reflection
UPDATE: Our api is not gRPC based; gRPC services are defined with a protobuf schema; gRPC supports a reflection API that allows other tools to consume the schema implemented by a server and generate various things, from dyanamic clients to API playgrounds or documentation generators.

Old goal was:

An HTTP GET to `/` should return a description of the API endpoints available. The debug output from the Router in HTTP routes should probably suffice as it's fairly descriptive and, I believe, JSON.

### PR DESCRIPTION
**Pull request by tustvold**
Closes #721

This is the simplest way to implement the _desired_ functionality but it is far from ideal, as the output is not JSON nor particularly easy to understand.

```
{ Pre-Middlewares: [], Routes: [{ path: "/ping/", regex: (?s)^/ping/$, route_params: [], methods: [GET] }, { path: "/api/v2/write/", regex: (?s)^/api/v2/write/$, route_params: [], methods: [POST] }, { path: "/api/v2/read/", regex: (?s)^/api/v2/read/$, route_params: [], methods: [GET] }, { path: "/iox/api/v1/databases/:name/", regex: (?s)^/iox/api/v1/databases/([^/]+)/$, route_params: ["name"], methods: [PUT] }, { path: "/iox/api/v1/databases/:name/", regex: (?s)^/iox/api/v1/databases/([^/]+)/$, route_params: ["name"], methods: [GET] }, { path: "/iox/api/v1/id/", regex: (?s)^/iox/api/v1/id/$, route_params: [], methods: [PUT] }, { path: "/api/v1/partitions/", regex: (?s)^/api/v1/partitions/$, route_params: [], methods: [GET] }, { path: "/api/v1/snapshot/", regex: (?s)^/api/v1/snapshot/$, route_params: [], methods: [POST] }], Post-Middlewares: [], ScopedDataMaps: [], ErrHandler: false, ShouldGenReqInfo: None }
```

Unless anyone has strong opinions to the contrary I'd be inclined to not merge this, and to instead create a custom API builder that not only can have a defined, readable and testable output format, but also provide an extension point for adding descriptions and potentially usage instructions as discussed on #729

### CODE REVIEW

File: src/influxdb_ioxd/http_routes.rs [hunk 1]
```diff
@@ -246,15 +262,16 @@ where
...
-        .get("/iox/api/v1/databases/:name", get_database::<M>)
-        .put("/iox/api/v1/id", set_writer::<M>)
-        .get("/api/v1/partitions", list_partitions::<M>)
-        .post("/api/v1/snapshot", snapshot_partition::<M>)
+        }))
+        .scope("", routes)
+        .get("/", move |_| {
+            let response = Response::builder()
+                .status(StatusCode::OK)
+                .body(Body::from(api_description.clone()))
```
Comment (alamb) [inline src/influxdb_ioxd/http_routes.rs:270]: I think outputting the `Debug` format of the routes, as you say, is pretty rough to grok

I was imagining the code looks something like this:

```
let routes: IOxRoutes = IoxRouterBuilder::new()
  .add(IoxRouteBuilder::new()
    .get("/ping", ping)
    .description("This route returns a PONG")
    .build())
  ...
  ).build();
```


And then there is something like
```
struct IOxRoutes {
    /// renders a human readable HTML description of all routes and how to use them
    pub fn html(&self) -> String {...}

    /// in my dream world (aka some future time)
    /// render the routes as a well formatted Swagger/OpenAPI spec
    pub fn open_api(&self) -> String {...}
...
}

**Comment by pauldix** [comment]
Yeah, that looks pretty rough. I'm all for doing something that is a bit better. For the documentation, I think there's some of it that would be put with where the routes are defined, but also some that would go with the Rust struct definitions for what gets read in and returned.

File: src/influxdb_ioxd/http_routes.rs (hunk 1 above)
Comment (tustvold) [reply src/influxdb_ioxd/http_routes.rs:270]: Yeah, something like this was what I was thinking of, potentially with an additional `.request<RequestType>` and `.response<ResponseType>` to cover what @pauldix refers to below

**Comment by tustvold** [comment]
There seems to be consensus not to move forward with this approach, so closing

**Pull request closed without merging**